src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java

```diff
@@ -0,0 +1,1727 @@
+        } else {
+          builder.field(fieldName, Schema.BOOLEAN_SCHEMA);
+        }
+        break;
+      }
+
+      // ints <= 8 bits
+      case Types.BIT: {
+        if (optional) {
+          builder.field(fieldName, Schema.OPTIONAL_INT8_SCHEMA);
+        } else {
+          builder.field(fieldName, Schema.INT8_SCHEMA);
+        }
+        break;
+      }
+
+      case Types.TINYINT: {
+        if (optional) {
+          if (columnDefn.isSignedNumber()) {
+            builder.field(fieldName, Schema.OPTIONAL_INT8_SCHEMA);
+          } else {
+            builder.field(fieldName, Schema.OPTIONAL_INT16_SCHEMA);
+          }
+        } else {
+          if (columnDefn.isSignedNumber()) {
+            builder.field(fieldName, Schema.INT8_SCHEMA);
+          } else {
+            builder.field(fieldName, Schema.INT16_SCHEMA);
+          }
+        }
+        break;
+      }
+
+      // 16 bit ints
+      case Types.SMALLINT: {
+        if (optional) {
+          if (columnDefn.isSignedNumber()) {
+            builder.field(fieldName, Schema.OPTIONAL_INT16_SCHEMA);
+          } else {
+            builder.field(fieldName, Schema.OPTIONAL_INT32_SCHEMA);
+          }
+        } else {
+          if (columnDefn.isSignedNumber()) {
+            builder.field(fieldName, Schema.INT16_SCHEMA);
+          } else {
+            builder.field(fieldName, Schema.INT32_SCHEMA);
+          }
+        }
+        break;
+      }
+
+      // 32 bit ints
+      case Types.INTEGER: {
+        if (optional) {
+          if (columnDefn.isSignedNumber()) {
+            builder.field(fieldName, Schema.OPTIONAL_INT32_SCHEMA);
+          } else {
+            builder.field(fieldName, Schema.OPTIONAL_INT64_SCHEMA);
+          }
+        } else {
+          if (columnDefn.isSignedNumber()) {
+            builder.field(fieldName, Schema.INT32_SCHEMA);
+          } else {
+            builder.field(fieldName, Schema.INT64_SCHEMA);
+          }
+        }
+        break;
+      }
+
+      // 64 bit ints
+      case Types.BIGINT: {
+        if (optional) {
+          builder.field(fieldName, Schema.OPTIONAL_INT64_SCHEMA);
+        } else {
+          builder.field(fieldName, Schema.INT64_SCHEMA);
+        }
+        break;
+      }
+
+      // REAL is a single precision floating point value, i.e. a Java float
+      case Types.REAL: {
+        if (optional) {
+          builder.field(fieldName, Schema.OPTIONAL_FLOAT32_SCHEMA);
+        } else {
+          builder.field(fieldName, Schema.FLOAT32_SCHEMA);
+        }
+        break;
+      }
+
+      // FLOAT is, confusingly, double precision and effectively the same as DOUBLE. See REAL
+      // for single precision
+      case Types.FLOAT:
+      case Types.DOUBLE: {
+        if (optional) {
+          builder.field(fieldName, Schema.OPTIONAL_FLOAT64_SCHEMA);
+        } else {
+          builder.field(fieldName, Schema.FLOAT64_SCHEMA);
+        }
+        break;
+      }
+
+      case Types.NUMERIC:
```

Comment:
@kkonstantine this is the first of two blocks which incorporate the substantiative portions of your recent changes from #396. Would love a sanity check here.